typegate/src/config/types.ts

```diff
@@ -32,21 +32,45 @@ export type GlobalConfig = z.infer<typeof globalConfigSchema>;
 // These config entries are only accessible on a Typegate instance.
 // They are not read from a global variable to enable test isolation and configurability.
 export const typegateConfigBaseSchema = z.object({
-  tg_secret: z.string().transform((s: string, ctx) => {
-    const bytes = decodeBase64(s);
-    if (bytes.length != 64) {
-      ctx.addIssue({
-        code: z.ZodIssueCode.custom,
-        message:
-          `Base64 contains ${bytes.length} instead of 64 bytes (use openssl rand -base64 64 | tr -d '\n')`,
-      });
-    }
-    return bytes;
-  }),
+  tg_secret: z
+    .string()
+    .optional()
+    .transform((s: string | undefined, ctx) => {
+      if (s === undefined) {
+        console.error(
+          "\x1b[31m%s\x1b[0m",
+          "Error: Env var TG_SECRET is not configured. Exiting from the application.",
+        );
+        Deno.exit(1);
+      }
+
+      const bytes = decodeBase64(s);
+      if (bytes.length != 64) {
+        ctx.addIssue({
+          code: z.ZodIssueCode.custom,
+          message:
+            `Base64 contains ${bytes.length} instead of 64 bytes (use openssl rand -base64 64 | tr -d '\n')`,
+        });
+      }
+      return bytes;
+    }),
   timer_max_timeout_ms: z.coerce.number().positive().max(60000),
   timer_destroy_resources: z.boolean(),
   timer_policy_eval_retries: z.number().nonnegative().max(5),
-  tg_admin_password: z.string(),
+  tg_admin_password: z
+    .string()
+    .optional()
+    .transform((s: string | undefined, _ctx) => {
+      if (s === undefined) {
+        console.error(
+          "\x1b[31m%s\x1b[0m",
+          "Error: Env var TG_ADMIN_PASSWORD is not configured. Exiting from the application.",
+        );
+        Deno.exit(1);
```

Comment:
I would advise against effects on validation, how about `ctx.addIssue({ ... })`?

Reply:
that works but the output on the console is kind of verbose(especially when there are multiple issues), I can change that to only show the issue on top of the stack and elevate the verbosity... that's why I chose `console.error()`, wdyt though? change how we display those Zod validation errors?